applications/StructuralMechanicsApplication/python_scripts/structural_response.py

```diff
@@ -220,7 +220,215 @@ def GetShapeGradient(self):
         return gradient
 
 # ==============================================================================
+class AdjointResponseFunction(ResponseFunctionBase):
+    """Linear static adjoint response function.
+    - runs the primal analysis (writes the primal results to an .h5 file)
+    - reads the primal results from the .h5 file into the adjoint model part
+    - uses primal results to calculate value
+    - uses primal results to calculate gradient by running the adjoint analysis
 
+    Attributes
+    ----------
+    primal_analysis : Primal analysis object of the response function
+    adjoint_analysis : Adjoint analysis object of the response function
+    response_function_utility: Cpp utilities object doing the actual computation of response value and gradient.
+    """
+    def __init__(self, identifier, project_parameters, model_part = None):
+        self.identifier = identifier
+
+        model = Model()
+        if model_part is not None:
+            model.AddModelPart(model_part)
+
+        # Create the primal solver
+        ProjectParametersPrimal = Parameters( open(project_parameters["primal_settings"].GetString(),'r').read() )
+        self.primal_analysis = structural_mechanics_analysis.StructuralMechanicsAnalysis(model, ProjectParametersPrimal)
+        self.primal_model_part_name = ProjectParametersPrimal["problem_data"]["model_part_name"].GetString()
+
+        # Create the adjoint solver
+        ProjectParametersAdjoint = Parameters( open(project_parameters["adjoint_settings"].GetString(),'r').read() )
+        ProjectParametersAdjoint["solver_settings"].AddValue("response_function_settings", project_parameters)
+
+        adjoint_model = Model()
```

Comment:
Who is owning the adjoint_model?

Reply:
Currently we need to create two models, one for the adjoint, one for the primal.
The problem is that currently the HDF5 process requires the adjoint model part to have the same name as the primal model part. So they cannot coexist in a single Model. This is why currently there is a second adjoint model.

As far as i understood, the Model should become a single global object in Kratos. So some adaption will be necessary.

For now: The model is stored inside the analysis object and is owned by it.

Reply:
ok